### PR TITLE
Mercadopago

### PR DESCRIPTION
**Pull request by dylan-muszel**
### Summary

Mercado Pago recipe.
Code was reviewed on: https://github.com/Wolox/wolmo-cookbook-android/pull/35, https://github.com/Wolox/wolmo-cookbook-android/pull/34, https://github.com/Wolox/wolmo-cookbook-android/pull/33.

### Screenshot

image

### CODE REVIEW

File: app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoItemsAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import ar.com.wolox.android.cookbook.R
+import ar.com.wolox.android.cookbook.databinding.ViewMercadopagoItemBinding
+import ar.com.wolox.android.cookbook.mercadopago.model.Product
+
+class MercadoPagoItemsAdapter(
+    private val items: MutableList<Pair<Product, Int>>
+) : RecyclerView.Adapter<MercadoPagoItemsAdapter.ViewHolder>() {
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoItemsAdapter.kt:12]: I'm trying to promote the use of `ListAdapter`. Check if you can extend from that class.

File: app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoItemsAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        ViewHolder(ViewMercadopagoItemBinding.inflate(LayoutInflater.from(parent.context), parent, false))
+
+    override fun getItemCount() = items.size
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(items[position])
+
+    class ViewHolder(private val binding: ViewMercadopagoItemBinding) : RecyclerView.ViewHolder(binding.root) {
+
+        fun bind(item: Pair<Product, Int>) {
+            binding.itemTitle.text = item.first.name
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoItemsAdapter.kt:24]: Use `with`

File: app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoRecipeFragment.kt [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        return binding.root
+    }
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+        _binding = null
+    }
+
+    override fun init() {
+        binding.payButton.setOnClickListener {
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoRecipeFragment.kt:70]: Use `with(binding)` here

Comment (lmiotti) [reply app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoRecipeFragment.kt:70]: See the whole class too

File: app/src/main/res/layout/fragment_mercadopago.xml [hunk 4]
```diff
@@ -0,0 +1,108 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (lmiotti) [inline app/src/main/res/layout/fragment_mercadopago.xml:2]: xmlns in new line

File: app/src/main/res/layout/view_mercadopago_item.xml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (lmiotti) [inline app/src/main/res/layout/view_mercadopago_item.xml:2]: same as above

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -246,4 +248,20 @@
...
+    <string name="mercadopago_pay_button">Let\'s pay!</string>
+    <string name="mercadopago_name_hint">Name</string>
+    <string name="mercadopago_item_total">\$%.2f x %d</string>
+    <string name="mercadopago_total_text">Total</string>
+    <string name="mercadopago_total_number">\$%.2f</string>
+    <string name="mercadopago_email_hint">Email</string>
+    <string name="mercadopago_empty_name_error">Please complete your name</string>
+    <string name="mercadopago_empty_email_error">Please complete your email</string>
+    <string name="mercadopago_invalid_email_error">Please complete a valid email</string>
+>>>>>>> Stashed changes
```
Comment (lmiotti) [inline app/src/main/res/values/strings.xml:265]: 👀

File: app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoItemsAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import ar.com.wolox.android.cookbook.R
+import ar.com.wolox.android.cookbook.databinding.ViewMercadopagoItemBinding
+import ar.com.wolox.android.cookbook.mercadopago.model.Product
+
+class MercadoPagoItemsAdapter(
+    private val items: MutableList<Pair<Product, Int>>
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoItemsAdapter.kt:11]: Why it's necessary to be mutable?

File: app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoRecipePresenter.kt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+class MercadoPagoRecipePresenter @Inject constructor(
+    @Named(CoroutineDispatchersModule.MAIN) mainDispatcher: CoroutineDispatcher,
+    private val mercadoPagoProductsRepository: MercadoPagoProductsRepository,
+    private val mercadoPagoAdapter: MercadoPagoAdapter
+) : CoroutineBasePresenter<MercadoPagoView>(mainDispatcher) {
+
+    private var items: List<Pair<Product, Int>>? = null
+
+    override fun onViewAttached() {
+        view?.startLoading()
```
Comment (lmiotti) [inline app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoRecipePresenter.kt:29]: Use `view?.run{}` instead

File: app/src/main/res/layout/fragment_mercadopago.xml (hunk 4 above)
Comment (dylan-muszel) [reply app/src/main/res/layout/fragment_mercadopago.xml:2]: oh, my reformat code put that there, but ok 🤷

File: app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoRecipePresenter.kt (hunk 8 above)
Comment (dylan-muszel) [reply app/src/main/java/ar/com/wolox/android/cookbook/mercadopago/MercadoPagoRecipePresenter.kt:29]: Using that will add more kind of callback hell for none reason. IMO doing that would make the code less readable.

**Pull request merged**